aip/general/0129.md

```diff
@@ -0,0 +1,71 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-modified Values and Defaults
+
+### State-driven clients (aka Convergent clients)
+Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.
+
+### Permadiff
+Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.
+
+## Guidance
+
+### Server Side Defaults
+There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.
+
+#### Dynamic Defaults
```

Comment:
I'd recommend calling these "Effective values" rather than "Dynamic Defaults".

As a metaphor, if we had prior art involving "Dynamic Constants" and "Static Constants", that would be misleading. A "static constant" is just a "constant", and a "dynamic constant" is just a "variable", and describing those situations as two kinds of constants is misleading/counterproductive.

In other words - there is no such thing as a "dynamic default". But, like you describe, there may be:
* a normal field and an effective value pair
* the normal field may allow a setting which means the server allocates or chooses the effective value
* the default value for the normal field may in fact be that setting where the effective value is expected to be allocated or algorithmic.

Reply:
I think that "dynamic default" describes the way these types of fields are handled today, where the default is determined to be a non-static value.

I'll change to effective values though, as the proposal is to split them into a normal field + output-only effective value

Reply:
Much appreciated! That's the only reason I'd feel strongly about terminology at the end of the day, 

When a client configured "", what that value "" meant is "dynamic" or "calculated" or "latest", and the service would overwrite the fact of that configuration with an effective value that resulted from an algorithm or allocation... Essentially a lossy change, causes a permadiff, but throws out the fact that the cluster was configured for "latest" rather than what actually took place.

Then we can reserve the term "dynamic default" for that existing anti-pattern, because that's how it's used today, and not need to clarify the difference between "old dynamic defaults" and "new dynamic defaults".

Reply:
We could always clarify it with "dynamic defaults expressed as effective fields" to avoid ambiguity.

Reply:
Changed to effective values in all places